Make sure TSan does not report false positives on IRGened code

AnnaZaks opened the pull request:
TSan does not see all of the synchronization happening in the runtime since the runtime is not instrumented. Due to tight interplay between the runtime and the generated code, we report false positives when metadata or witness tables are accessed on certain types.

Fix these false positives by modifying the IRGen to hide writes in some cases and exposing stronger barriers in others.

Additionally, now that we have compiler-rt, add basic sanitizer tests.

@swift-ci Please test

AnnaZaks commented:
@swift-ci Please test

gribozavr commented on test/Sanitizers/witness_table_lookup.swift:
```diff
@@ -0,0 +1,64 @@
+}
+
+func testForRaces() {
+  let queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)
```
Don't race tests from StdlibUnittest work for you?

AnnaZaks commented:
@swift-ci Please test

AnnaZaks commented:
@rjmccall, 
Please review. Thanks!

rjmccall commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -1224,7 +1224,10 @@ createInPlaceMetadataInitializationFunction(IRGenModule &IGM,
+  // Skip instrumentation when building for TSan to avoid false positives.
+  // The syncronization for this happens in the Runtime and we do not see it.
+  if (IGM.IRGen.Opts.Sanitize == SanitizerKind::Thread)
+    fn->removeFnAttr(llvm::Attribute::SanitizeThread);
```
Please keep a blank line after this, thanks.

rjmccall commented:
Minor style nit-pick, and I'll let Dmitri's question stand about testing, but otherwise this LGTM.

AnnaZaks commented:
Addressed all review comments.
@swift-ci Please test

gribozavr commented on test/Sanitizers/tsan-type-metadata.swift:
```diff
@@ -0,0 +1,96 @@
+// We have to make sure TSan does not see the acesses to the metadata from the IR.
+// Otherwise, it will report a race.
+
+import Foundation
```
Do we need to import Foundation?  Seems like we are not using it for anything.

AnnaZaks replied:
Extending from NSObject:

AnnaZaks replied:
private class KeyWrapper<T: Hashable>: NSObject

gribozavr replied:
Is this important for the test?  Can we drop it?

AnnaZaks commented:
- Removed dependency on NSObject
- Updated the simple sanitizer tests so that they do not crash to avoid generating crash reports.
- Updated the metadata and witness tests to stop on first error to ensure that StdlibUnittest errors out of a race is reported.

@swift-ci Please test

AnnaZaks commented:
@swift-ci Please test

AnnaZaks commented:
Foundation failures are unrelated.
